src/modules/lighthouse/Eth2Client.ts

```diff
@@ -331,73 +349,41 @@ export class Eth2Client {
     return ((await this.getGenesisTime()) + (slot * BigInt(this.environment.CHAIN_SLOT_TIME_SECONDS)));
   }
 
-  protected get url(): string {
-    // sanity check
-    if (this.activeRpcUrl > this.rpcUrls.length - 1) {
-      this.activeRpcUrl = 0;
-    }
-
-    return this.rpcUrls[this.activeRpcUrl];
-  }
-
-  protected switchToNextRpc(): void {
-    if (this.rpcUrls.length === 1) {
-      this.logger.info('Will not switch to next RPC url for ETH2. No backup RPC provided.');
-      return;
-    }
-    this.activeRpcUrl++;
-    this.logger.info('Switched to next RPC url for ETH2');
-  }
-
-  protected async retryGet<T>(subUrl: string, maxRetries = 5, getFunc: GetFunction = this.apiGetNoFallback, dataOnly = true): Promise<T> {
-    const retry = retrier(this.logger, maxRetries, 100, 10000, true);
-
-    const res = await getFunc(subUrl)
-      .catch(rejectDelay(this.environment.ETH2_BEACON_RPC_RETRY_DELAY_MS))
-      .catch(() => retry(() => getFunc(subUrl)))
-      .catch((e) => {
-        this.logger.error('Error while doing ETH2 RPC request. Will try to switch to another RPC');
-        this.switchToNextRpc();
-        throw e;
-      })
-      .catch(() => retry(() => getFunc(subUrl)));
-
-    if (dataOnly) return res.data; else return res;
-  }
-
-  protected async retryPost<T>(
-    subUrl: string, maxRetries = 5, params?: Record<string, unknown>, postFunc: PostFunction = this.apiPostNoFallback, dataOnly = true
-  ): Promise<T> {
+  protected async retryRequest<T>(callback: (rpcURL: string) => any, options?: RequestRetryOptions): Promise<T> {
+    const [maxRetries, dataOnly, fallbackConditionCallback] = [
+      options?.maxRetries ?? 5,
+      options?.dataOnly != undefined ? options.dataOnly : true,
+      options?.fallbackConditionCallback != undefined ? options.fallbackConditionCallback : () => true
+    ]
     const retry = retrier(this.logger, maxRetries, 100, 10000, true);
-
-    const res = await postFunc(subUrl, params)
+    const res = await callback(this.rpcUrls.main)
       .catch(rejectDelay(this.environment.ETH2_BEACON_RPC_RETRY_DELAY_MS))
-      .catch(() => retry(() => postFunc(subUrl, params)))
-      .catch((e) => {
-        this.logger.error('Error while doing ETH2 RPC request. Will try to switch to another RPC');
-        this.switchToNextRpc();
+      .catch(() => retry(() => callback(this.rpcUrls.main)))
+      .catch((e: any) => {
+        if (fallbackConditionCallback(e)) {
+          this.logger.error('Error while doing ETH2 RPC request. Will try to switch to another RPC if exists');
+          return retry(() => callback(this.rpcUrls.backup || this.rpcUrls.main));
```

Comment:
May be it is better to basically throw an error if not backup url provided?

Reply:
It seems to me, a good option would be to write a warning and not try to use backup url

Reply:
fixed